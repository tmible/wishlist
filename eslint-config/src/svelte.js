import svelte from "eslint-plugin-svelte";
import globals from "globals";

export default [
  ...svelte.configs["flat/recommended"],
  {
    languageOptions: {
      globals: {
        ...globals.browser,
        ...globals.node,
      },
      ecmaVersion: "latest",
      sourceType: "module",
      parserOptions: {
        extraFileExtensions: [ ".svelte" ],
      },
    },

    settings: {
      "import/parsers": {
        "svelte-eslint-parser": [ ".svelte" ],
        "espree": [ ".js" ],
      },
    },

    rules: {
      /* import */
      // svelte does tree shaking when compiling
      "import/no-namespace": "off",

      /* svelte */
      "svelte/infinite-reactive-loop": "error",
      "svelte/no-dom-manipulating": "error",
      "svelte/no-dupe-on-directives": "error",
      "svelte/no-dupe-use-directives": "error",
      "svelte/no-reactive-reassign": "error",
      "svelte/no-store-async": "error",
      "svelte/require-store-callbacks-use-set-param": "error",
      "svelte/block-lang": "error",
      "svelte/no-ignored-unsubscribe": "error",
      "svelte/no-immutable-reactive-statements": "error",
      "svelte/no-inline-styles": "error",
      "svelte/no-reactive-functions": "error",
      "svelte/no-reactive-literals": "error",
      "svelte/no-useless-mustaches": "error",
      "svelte/require-each-key": "error",
      "svelte/require-event-dispatcher-types": "error",
      "svelte/require-stores-init": "error",
      "svelte/valid-each-key": "error",
      "svelte/derived-has-same-inputs-outputs": "error",
      "svelte/first-attribute-linebreak": "error",
      "svelte/html-closing-bracket-new-line": "error",
      "svelte/html-closing-bracket-spacing": "error",
      "svelte/html-quotes": "error",
      "svelte/html-self-closing": [
        "error",
        {
          "void": "never",
          "normal": "never",
          "svg": "always",
          "math": "never",
          "component": "always",
          "svelte": "always",
        },
      ],
      "svelte/indent": [ "error", { ignoredNodes: [ "ConditionalExpression" ] } ],
      "svelte/max-attributes-per-line": [ "error", { multiline: 1, singleline: 100 } ],
      "svelte/mustache-spacing": "error",
      "svelte/no-extra-reactive-curlies": "error",
      "svelte/no-spaces-around-equal-signs-in-attribute": "error",
      "svelte/prefer-class-directive": "error",
      "svelte/prefer-style-directive": "error",
      "svelte/shorthand-attribute": "error",
      "svelte/sort-attributes": [
        "error",
        {
          order: [
            // `this` property.
            "this",
            // `bind:this` directive.
            "bind:this",
            // `id` attribute.
            "id",
            // `name` attribute.
            "name",
            // `slot` attribute.
            "slot",
            // `--style-props`
            { match: "/^--/u", sort: "ignore" },
            // `style` attribute, and `style:` directives.
            [ "style", "/^style:/u" ],
            // `class` attribute.
            "class",
            // `class:` directives.
            { match: "/^class:/u", sort: "ignore" },
            // other attributes.
            {
              match: [ "!/:/u", "!/^(?:this|id|name|style|class)$/u", "!/^--/u" ],
              sort: "ignore",
            },
            // `bind:` directives (other then `bind:this`), and `on:` directives.
            [ "/^bind:/u", "!bind:this", "/^on:/u" ],
            // `use:` directives.
            { match: "/^use:/u", sort: "ignore" },
            // `transition:` directive.
            { match: "/^transition:/u", sort: "ignore" },
            // `in:` directive.
            { match: "/^in:/u", sort: "ignore" },
            // `out:` directive.
            { match: "/^out:/u", sort: "ignore" },
            // `animate:` directive.
            { match: "/^animate:/u", sort: "ignore" },
            // `let:` directives.
            { match: "/^let:/u", sort: "ignore" },
          ],
        },
      ],
      "svelte/spaced-html-comment": "error",
      "svelte/no-inner-declarations": "error",
      "svelte/no-trailing-spaces": "error",
    },
  },
  {
    files: [ "**/*.svelte" ],
    rules: {
      // svelte components can import themselves
      "import/no-self-import": "off",
    },
  },
  {
    files: [ "**/**.test.js", "**/**.test.svelte.js", "**/__tests__/**" ],
    settings: { "import/ignore": [ "vitest", "@testing-library/svelte" ] },
  },
];
