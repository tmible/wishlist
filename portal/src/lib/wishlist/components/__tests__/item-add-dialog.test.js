// @vitest-environment jsdom
import { cleanup, render, screen } from '@testing-library/svelte';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import ItemAddDialog from '../item-add-dialog.svelte';

let formMock = null;

vi.mock('../item-form.svelte', () => ({ default: (_, props) => formMock = props }));

describe('wishlist / components / item add dialog', () => {
  let modalPortal;

  beforeEach(() => {
    modalPortal = document.createElement('div');
    modalPortal.id = 'modal-portal';
    document.body.append(modalPortal);
  });

  afterEach(() => {
    formMock = null;
    modalPortal.remove();
    vi.clearAllMocks();
    cleanup();
  });

  it('should be displayed closed', () => {
    const { baseElement } = render(ItemAddDialog, { open: false });
    expect(baseElement).toMatchSnapshot();
  });

  describe('opened', () => {
    let baseElement;

    beforeEach(() => {
      vi.useFakeTimers();
      ({ baseElement } = render(ItemAddDialog, { open: true }));
    });

    afterEach(() => {
      vi.clearAllTimers();
      vi.useRealTimers();
    });

    it('should be displayed', () => {
      expect(baseElement).toMatchSnapshot();
    });

    it('should focus itself', () => {
      const dialog = screen.getByRole('dialog');
      vi.spyOn(dialog, 'focus');
      vi.runAllTimers();
      expect(vi.mocked(dialog.focus)).toHaveBeenCalled();
    });

    it('should close on form finish', () => {
      formMock.onfinish();
      expect(baseElement).toMatchSnapshot();
    });
  });
});
