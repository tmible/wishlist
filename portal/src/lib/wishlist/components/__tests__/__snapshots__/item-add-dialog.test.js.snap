// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wishlist / components / item add dialog > opened > should be displayed 1`] = `
<body
  style="margin-right: 8px; --scrollbar-width: 1024px; overflow: hidden; pointer-events: none;"
>
  <div
    id="modal-portal"
  >
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <div
      class="modal-backdrop"
      data-dialog-overlay=""
      data-state="open"
      id="bits-5"
      style="pointer-events: auto;"
    >
      <!---->
    </div>
    
    
    
    <!---->
     
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
     
    <div
      aria-labelledby="bits-12"
      aria-modal="true"
      class="modal-box overflow-y-visible p-0"
      data-dialog-content=""
      data-state="open"
      id="bits-6"
      interactoutsidebehavior="close"
      role="dialog"
      style="pointer-events: auto; outline: undefined;"
      tabindex="-1"
    >
      <!---->
      <!---->
      <div
        class="max-h-[inherit]"
        data-scroll-area-root=""
        dir="ltr"
        id="bits-9"
        style="position: relative; --bits-scroll-area-corner-height: 0px; --bits-scroll-area-corner-width: 0px;"
      >
        <!---->
        <div
          class="max-h-[inherit] p-6"
          data-scroll-area-viewport=""
          id="bits-10"
          style="overflow-x: hidden; overflow-y: scroll;"
        >
          <div
            data-scroll-area-content=""
            id="bits-11"
            style="min-width: undefined;"
          >
            <!---->
            <div
              class="flex items-start prose"
            >
              <!---->
              <div
                aria-level="2"
                class="mt-0 mr-auto"
                data-dialog-title=""
                data-state="open"
                id="bits-12"
                role="heading"
              >
                Добавить желание
                <!---->
              </div>
              
              <!---->
               
              <!---->
              <button
                class="cursor-pointer"
                data-dialog-close=""
                data-state="open"
                id="bits-13"
                tabindex="0"
              >
                <svg
                  class="lucide-icon lucide lucide-x"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <!---->
                  <path
                    d="M18 6 6 18"
                  >
                    
                  </path>
                  
                  <!---->
                  <path
                    d="m6 6 12 12"
                  >
                    
                  </path>
                  
                  <!---->
                  <!---->
                  
                  <!---->
                </svg>
                <!---->
              </button>
              
              <!---->
            </div>
             
            <!---->
            
            <!---->
          </div>
        </div>
        <!---->
         
        <!---->
        <!---->
        <!---->
        
        
        
        <!---->
         
        <!---->
         
        <!---->
        
        <!---->
        <!---->
      </div>
      
      
      <!---->
    </div>
    
    
    
    
    
    
    
    <!---->
    
    
    
  </div>
  <div>
    <!---->
    <!---->
    <!---->
    <!---->
    <button
      aria-controls="bits-6"
      aria-expanded="true"
      aria-haspopup="dialog"
      class="hidden"
      data-dialog-trigger=""
      data-state="open"
      id="bits-4"
    >
      <!---->
    </button>
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
</body>
`;

exports[`wishlist / components / item add dialog > opened > should close on form finish 1`] = `
<body
  style="margin-right: 8px; --scrollbar-width: 1024px; overflow: hidden; pointer-events: none;"
>
  <div
    id="modal-portal"
  >
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <div
      class="modal-backdrop"
      data-dialog-overlay=""
      data-state="open"
      id="bits-29"
      style="pointer-events: auto;"
    >
      <!---->
    </div>
    
    
    
    <!---->
     
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
     
    <div
      aria-labelledby="bits-36"
      aria-modal="true"
      class="modal-box overflow-y-visible p-0"
      data-dialog-content=""
      data-state="open"
      id="bits-30"
      interactoutsidebehavior="close"
      role="dialog"
      style="pointer-events: auto; outline: undefined;"
      tabindex="-1"
    >
      <!---->
      <!---->
      <div
        class="max-h-[inherit]"
        data-scroll-area-root=""
        dir="ltr"
        id="bits-33"
        style="position: relative; --bits-scroll-area-corner-height: 0px; --bits-scroll-area-corner-width: 0px;"
      >
        <!---->
        <div
          class="max-h-[inherit] p-6"
          data-scroll-area-viewport=""
          id="bits-34"
          style="overflow-x: hidden; overflow-y: scroll;"
        >
          <div
            data-scroll-area-content=""
            id="bits-35"
            style="min-width: undefined;"
          >
            <!---->
            <div
              class="flex items-start prose"
            >
              <!---->
              <div
                aria-level="2"
                class="mt-0 mr-auto"
                data-dialog-title=""
                data-state="open"
                id="bits-36"
                role="heading"
              >
                Добавить желание
                <!---->
              </div>
              
              <!---->
               
              <!---->
              <button
                class="cursor-pointer"
                data-dialog-close=""
                data-state="open"
                id="bits-37"
                tabindex="0"
              >
                <svg
                  class="lucide-icon lucide lucide-x"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <!---->
                  <path
                    d="M18 6 6 18"
                  >
                    
                  </path>
                  
                  <!---->
                  <path
                    d="m6 6 12 12"
                  >
                    
                  </path>
                  
                  <!---->
                  <!---->
                  
                  <!---->
                </svg>
                <!---->
              </button>
              
              <!---->
            </div>
             
            <!---->
            
            <!---->
          </div>
        </div>
        <!---->
         
        <!---->
        <!---->
        <!---->
        
        
        
        <!---->
         
        <!---->
         
        <!---->
        
        <!---->
        <!---->
      </div>
      
      
      <!---->
    </div>
    
    
    
    
    
    
    
    <!---->
    
    
    
  </div>
  <div>
    <!---->
    <!---->
    <!---->
    <!---->
    <button
      aria-controls="bits-30"
      aria-expanded="true"
      aria-haspopup="dialog"
      class="hidden"
      data-dialog-trigger=""
      data-state="open"
      id="bits-28"
    >
      <!---->
    </button>
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
</body>
`;

exports[`wishlist / components / item add dialog > should be displayed closed 1`] = `
<body>
  <div
    id="modal-portal"
  >
    <!---->
    <!---->
    <!---->
    <!---->
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
  <div>
    <!---->
    <!---->
    <!---->
    <!---->
    <button
      aria-expanded="false"
      aria-haspopup="dialog"
      class="hidden"
      data-dialog-trigger=""
      data-state="closed"
      id="bits-1"
    >
      <!---->
    </button>
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
</body>
`;
